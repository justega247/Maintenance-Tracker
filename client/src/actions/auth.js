import axios from 'axios';
import Cookie from 'cookies-js';
import jwt from 'jsonwebtoken';
import {
  SET_CURRENT_USER,
  SET_CURRENT_USER_FAIL,
  LOGOUT_USER,
} from '../constants/actionTypes';
import config from '../config';
import { backendRoutes } from '../constants/routes';

// USER_LOGIN
export const setCurrentUser = user => ({
  type: SET_CURRENT_USER,
  user,
});

export const setCurrentUserError = error => ({
  type: SET_CURRENT_USER_FAIL,
  error,
});

export const startUserSignIn = ({
  username,
  password,
}) => (dispatch) => {
  const userDetails = { username, password };
  return axios
    .post(`${config.apiUrl}${backendRoutes.LOGIN}`, userDetails)
    .then((response) => {
      const { token } = response.data.data.user;
      Cookie.set('jwtToken', token);
      dispatch(setCurrentUser(jwt.decode(token)));
      return response;
    })
    .catch((error) => {
      const { message } = error.response.data;
      dispatch(setCurrentUserError(message));
    });
};

export const startUserSignUp = ({
  username, fullname, email, password,
}) => (dispatch) => {
  const userDetails = {
    username, fullname, email, password,
  };
  return axios.post(`${config.apiUrl}${backendRoutes.SIGN_UP}`, userDetails)
    .then((response) => {
      const { token } = response.data.data.user;
      Cookie.set('jwtToken', token);
      dispatch(setCurrentUser(jwt.decode(token)));
    })
    .catch((error) => {
      const { message } = error.response.data;
      dispatch(setCurrentUserError(message));
    });
};

export const logoutCurrentUser = () => ({
  type: LOGOUT_USER,
});

export const logoutAction = () => (dispatch) => {
  Cookie.expire('jwtToken');
  dispatch(logoutCurrentUser({}));
};
