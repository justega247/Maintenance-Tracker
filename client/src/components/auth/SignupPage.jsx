import React, { Component } from "react";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import Header from "../Header";
import Button from "../common/Button";
import ValidateUser from "../../utils/authValidation";
import routes from "../../constants/routes";

class SignupPage extends Component {
  state = {
    username: "",
    fullname: "",
    email: "",
    password: "",
    errors: {}
  };

  onChange = event => {
    const { name, value } = event.target;
    this.setState(() => ({
      [name]: value
    }));
  };

  onSubmit = event => {
    event.preventDefault();

    const { errors, hasErrors } = ValidateUser.signUpDataValidation(this.state);
    this.setState(() => ({
      errors
    }));

    if (!hasErrors) {
      const { username, fullname, email, password } = this.state;
      this.props.startUserSignUp({ username, fullname, email, password });
    }
  };

  render() {
    const { errors } = this.state;

    return (
      <div>
        <Header />
        <form onSubmit={this.onSubmit} className="auth__form">
          <div className="auth__body">
            <div>
              <label htmlFor="username">
                Username:
                <div>
                  <input
                    type="text"
                    name="username"
                    placeholder="enter your username"
                    className="form__input"
                    value={this.state.username}
                    onChange={this.onChange}
                  />
                </div>
              </label>
              {errors.username &&
                errors.username.map((error, index) => (
                  <div className="form__error" key={index}>
                    {error}
                  </div>
                ))}
            </div>
            <div>
              <label htmlFor="fullname">
                Fullname:
                <div>
                  <input
                    type="text"
                    name="fullname"
                    placeholder="enter your fullname"
                    className="form__input"
                    value={this.state.fullname}
                    onChange={this.onChange}
                  />
                </div>
              </label>
              {errors.fullname &&
                errors.fullname.map((error, index) => (
                  <div className="form__error" key={index}>
                    {error}
                  </div>
                ))}
            </div>
            <div>
              <label htmlFor="email">
                Email:
                <div>
                  <input
                    type="email"
                    name="email"
                    placeholder="enter your email"
                    className="form__input"
                    value={this.state.email}
                    onChange={this.onChange}
                  />
                </div>
              </label>
              {errors.email &&
                errors.email.map((error, index) => (
                  <div className="form__error" key={index}>
                    {error}
                  </div>
                ))}
            </div>
            <div>
              <label htmlFor="password">
                Password:
                <div>
                  <input
                    type="password"
                    name="password"
                    placeholder="enter your password"
                    className="form__input"
                    value={this.state.password}
                    onChange={this.onChange}
                  />
                </div>
              </label>
              {errors.password &&
                errors.password.map((error, index) => (
                  <div className="form__error" key={index}>
                    {error}
                  </div>
                ))}
            </div>
            <div>
              <Button className="button--big" type="submit" text="Signup" />
            </div>
            <div>
              <h4>
                Already have an account?{" "}
                <Link className="auth__extra-item" to={routes.SIGN_IN}>
                  Login
                </Link>
              </h4>
            </div>
          </div>
        </form>
      </div>
    );
  }
}

SignupPage.defaultProps = {
  startUserSignUp: () => {}
};

SignupPage.propTypes = {
  startUserSignUp: PropTypes.func
};

export default SignupPage;
