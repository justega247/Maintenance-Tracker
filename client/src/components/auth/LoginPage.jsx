import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import Header from '../Header';
import Button from '../common/Button';
import ValidateUser from '../../utils/authValidation';
import { frontendRoutes } from '../../constants/routes';
import { startUserSignIn } from '../../actions/auth';

export class LoginPage extends Component {
  state = {
    username: '',
    password: '',
    errors: {},
  };

  onChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };


  onSubmit = (event) => {
    event.preventDefault();

    const { errors, hasErrors } = ValidateUser.signInDataValidation(this.state);
    this.setState(() => ({
      errors,
    }));

    if (!hasErrors) {
      const { username, password } = this.state;
      const { startUserLogin, history } = this.props;
      startUserLogin({ username, password }, history);
    }
  };

  render() {
    const { errors } = this.state;

    return (
      <div>
        <Header />
        <div>
          <form onSubmit={this.onSubmit} className="auth__form">
            <div className="auth__body">
              <div>
                <label htmlFor="username" className="label">
                  Username:
                  <div>
                    <input
                      type="text"
                      name="username"
                      placeholder="enter your username"
                      className="form__input"
                      value={this.state.username}
                      onChange={this.onChange}
                    />
                  </div>
                </label>
                {errors.username &&
                  errors.username.map(error => (
                    <div className="form__error" key={error}>
                      {error}
                    </div>
                  ))}
              </div>
              <div>
                <label htmlFor="password" className="label">
                  Password:
                  <div>
                    <input
                      type="password"
                      name="password"
                      placeholder="enter your password"
                      className="form__input"
                      value={this.state.password}
                      onChange={this.onChange}
                    />
                  </div>
                </label>
                {errors.password &&
                  errors.password.map(error => (
                    <div className="form__error" key={error}>
                      {error}
                    </div>
                  ))}
              </div>
              <div>
                <Button className="button--big" type="submit" text="Login" />
              </div>
              <div className="auth__extra">
                <div className="auth__extra-item">
                  Yet to create an account?
                </div>
                <Link className="auth__extra-item" to={frontendRoutes.SIGN_UP}>
                  Sign up
                </Link>
              </div>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

LoginPage.defaultProps = {
  startUserLogin: () => {},
};

LoginPage.propTypes = {
  startUserLogin: PropTypes.func,
  history: PropTypes.instanceOf(Object).isRequired,
};

const mapStateToProps = state => ({
  auth: state.auth.isAuthenticated,
  error: state.auth.error,
});

const mapDispatchToProps = dispatch => ({
  startUserLogin: bindActionCreators(startUserSignIn, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(LoginPage);
