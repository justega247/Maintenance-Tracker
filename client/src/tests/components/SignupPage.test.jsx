import React from 'react';
import { shallow } from 'enzyme';
import { SignupPage } from '../../components/auth/SignupPage';

let startUserRegister;
let wrapper;
let history;

describe('Signup Page', () => {
  beforeEach(() => {
    startUserRegister = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<SignupPage
      startUserRegister={startUserRegister}
      history={history}
    />);
  });

  test('should correctly render SignupPage', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should set username on input change', () => {
    const value = 'tommy';
    const name = 'username';
    wrapper.find('input').at(0).simulate('change', {
      target: { name, value },
    });
    expect(wrapper.state(name)).toBe(value);
  });

  test('should set fullname on input change', () => {
    const value = 'tommy joe';
    const name = 'fullname';
    wrapper.find('input').at(1).simulate('change', {
      target: { name, value },
    });
    expect(wrapper.state(name)).toBe(value);
  });

  test('should set email on input change', () => {
    const value = 'tommy@gmail.com';
    const name = 'email';
    wrapper.find('input').at(2).simulate('change', {
      target: { name, value },
    });
    expect(wrapper.state(name)).toBe(value);
  });

  test('should set password on input change', () => {
    const value = 'tommyjoe';
    const name = 'password';
    wrapper.find('input').at(3).simulate('change', {
      target: { name, value },
    });
    expect(wrapper.state(name)).toBe(value);
  });

  test('should return error for invalid form submission', () => {
    wrapper.find('form').simulate('submit', {
      preventDefault: () => { },
    });
    expect(wrapper.state('errors')).not.toBe({});
    expect(wrapper).toMatchSnapshot();
  });

  test('should call startUserSignUp props for valid form submission', () => {
    const validDetails = {
      username: 'allison',
      fullname: 'alli paterson',
      email: 'all2@gmail.com',
      password: 'passing',
    };
    wrapper.setState(validDetails);
    wrapper.find('form').simulate('submit', {
      preventDefault: () => { },
    });
    expect(startUserRegister).toHaveBeenCalledWith(validDetails, history);
    expect(wrapper).toMatchSnapshot();
  });
});
