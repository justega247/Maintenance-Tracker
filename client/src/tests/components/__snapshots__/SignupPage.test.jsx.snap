// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Page should call startUserSignUp props for valid form submission 1`] = `
<div>
  <Header />
  <form
    className="auth__form"
    onSubmit={[Function]}
  >
    <div
      className="auth__body"
    >
      <div>
        <label
          className="label"
          htmlFor="username"
        >
          Username:
          <div>
            <input
              className="form__input"
              name="username"
              onChange={[Function]}
              placeholder="enter your username"
              type="text"
              value="allison"
            />
          </div>
        </label>
      </div>
      <div>
        <label
          className="label"
          htmlFor="fullname"
        >
          Fullname:
          <div>
            <input
              className="form__input"
              name="fullname"
              onChange={[Function]}
              placeholder="enter your fullname"
              type="text"
              value="alli paterson"
            />
          </div>
        </label>
      </div>
      <div>
        <label
          className="label"
          htmlFor="email"
        >
          Email:
          <div>
            <input
              className="form__input"
              name="email"
              onChange={[Function]}
              placeholder="enter your email"
              type="email"
              value="all2@gmail.com"
            />
          </div>
        </label>
      </div>
      <div>
        <label
          className="label"
          htmlFor="password"
        >
          Password:
          <div>
            <input
              className="form__input"
              name="password"
              onChange={[Function]}
              placeholder="enter your password"
              type="password"
              value="passing"
            />
          </div>
        </label>
      </div>
      <div>
        <Button
          className="button--big"
          text="Signup"
          type="submit"
        />
      </div>
      <div>
        <h4>
          Already have an account?
           
          <Link
            className="auth__extra-item"
            replace={false}
            to="/login"
          >
            Login
          </Link>
        </h4>
      </div>
    </div>
  </form>
</div>
`;

exports[`Signup Page should correctly render SignupPage 1`] = `
<div>
  <Header />
  <form
    className="auth__form"
    onSubmit={[Function]}
  >
    <div
      className="auth__body"
    >
      <div>
        <label
          className="label"
          htmlFor="username"
        >
          Username:
          <div>
            <input
              className="form__input"
              name="username"
              onChange={[Function]}
              placeholder="enter your username"
              type="text"
              value=""
            />
          </div>
        </label>
      </div>
      <div>
        <label
          className="label"
          htmlFor="fullname"
        >
          Fullname:
          <div>
            <input
              className="form__input"
              name="fullname"
              onChange={[Function]}
              placeholder="enter your fullname"
              type="text"
              value=""
            />
          </div>
        </label>
      </div>
      <div>
        <label
          className="label"
          htmlFor="email"
        >
          Email:
          <div>
            <input
              className="form__input"
              name="email"
              onChange={[Function]}
              placeholder="enter your email"
              type="email"
              value=""
            />
          </div>
        </label>
      </div>
      <div>
        <label
          className="label"
          htmlFor="password"
        >
          Password:
          <div>
            <input
              className="form__input"
              name="password"
              onChange={[Function]}
              placeholder="enter your password"
              type="password"
              value=""
            />
          </div>
        </label>
      </div>
      <div>
        <Button
          className="button--big"
          text="Signup"
          type="submit"
        />
      </div>
      <div>
        <h4>
          Already have an account?
           
          <Link
            className="auth__extra-item"
            replace={false}
            to="/login"
          >
            Login
          </Link>
        </h4>
      </div>
    </div>
  </form>
</div>
`;

exports[`Signup Page should return error for invalid form submission 1`] = `
<div>
  <Header />
  <form
    className="auth__form"
    onSubmit={[Function]}
  >
    <div
      className="auth__body"
    >
      <div>
        <label
          className="label"
          htmlFor="username"
        >
          Username:
          <div>
            <input
              className="form__input"
              name="username"
              onChange={[Function]}
              placeholder="enter your username"
              type="text"
              value=""
            />
          </div>
        </label>
        <div
          className="form__error"
          key="username field is required."
        >
          username field is required.
        </div>
      </div>
      <div>
        <label
          className="label"
          htmlFor="fullname"
        >
          Fullname:
          <div>
            <input
              className="form__input"
              name="fullname"
              onChange={[Function]}
              placeholder="enter your fullname"
              type="text"
              value=""
            />
          </div>
        </label>
        <div
          className="form__error"
          key="fullname is required."
        >
          fullname is required.
        </div>
      </div>
      <div>
        <label
          className="label"
          htmlFor="email"
        >
          Email:
          <div>
            <input
              className="form__input"
              name="email"
              onChange={[Function]}
              placeholder="enter your email"
              type="email"
              value=""
            />
          </div>
        </label>
        <div
          className="form__error"
          key="Please you have to specify a valid email so we can contact you."
        >
          Please you have to specify a valid email so we can contact you.
        </div>
      </div>
      <div>
        <label
          className="label"
          htmlFor="password"
        >
          Password:
          <div>
            <input
              className="form__input"
              name="password"
              onChange={[Function]}
              placeholder="enter your password"
              type="password"
              value=""
            />
          </div>
        </label>
        <div
          className="form__error"
          key="password field is required."
        >
          password field is required.
        </div>
      </div>
      <div>
        <Button
          className="button--big"
          text="Signup"
          type="submit"
        />
      </div>
      <div>
        <h4>
          Already have an account?
           
          <Link
            className="auth__extra-item"
            replace={false}
            to="/login"
          >
            Login
          </Link>
        </h4>
      </div>
    </div>
  </form>
</div>
`;
